src/Helpers/JWKFetcher.php

```diff
@@ -38,45 +45,115 @@ public function __construct(CacheHandler $cache = null, $guzzleOptions = [])
+     * @param string      $path   Path to the JWKS from the $domain.
      *
-     * @return array|mixed|null
+     * @return mixed
      *
-     * @throws \Exception
+     * @throws \Exception If the Guzzle HTTP client cannot complete the request.
      */
-    public function fetchKeys($iss)
+    public function fetchKeys($domain, $kid = null, $path = '.well-known/jwks.json')
     {
-        $url = "{$iss}.well-known/jwks.json";
+        $jwks_url = $domain.$path;
+
+        // Check for a cached JWKS value.
+        $secret = $this->cache->get($jwks_url);
+        if (! is_null($secret)) {
+            return $secret;
+        }
+
+        $secret = [];
+
+        $jwks = $this->getJwks( $domain, $path );
+
+        if (! is_array( $jwks['keys'] ) || empty( $jwks['keys'] )) {
+            return $secret;
+        }
 
-        if (($secret = $this->cache->get($url)) === null) {
-            $secret = [];
+        // No kid passed so get the kid of the first JWK.
+        if (is_null($kid)) {
+            $kid = $this->getProp( $jwks, 'kid' );
+        }
+
+        $x5c = $this->getProp( $jwks, 'x5c', $kid );
 
-            $request = new RequestBuilder([
-                'domain' => $iss,
-                'basePath' => '.well-known/jwks.json',
-                'method' => 'GET',
-                'guzzleOptions' => $this->guzzleOptions
-            ]);
-            $jwks    = $request->call();
+        // Need the kid and x5c for a well-formed return value. 
+        if (! is_null($kid) && ! is_null($x5c)) {
+            $secret[$kid] = $this->convertCertToPem($x5c);
+            $this->cache->set($jwks_url, $secret);
+        }
 
+        return $secret;
+    }
+
+    /**
+     * Get a specific property from a JWKS using a key, if provided.
+     *
+     * @param array       $jwks JWKS to parse.
+     * @param string      $prop Property to retrieve.
+     * @param null|string $kid  Kid to check.
+     *
+     * @return null|string
+     */
+    public function getProp(array $jwks, $prop, $kid = null)
+    {
+        $r_key = null;
+        if (! $kid) {
+            // No kid indicated, get the first entry.
+            $r_key = $jwks['keys'][0];
+        } else {
+            // Iterate through the JWKS for the correct kid.
             foreach ($jwks['keys'] as $key) {
-                $secret[$key['kid']] = $this->convertCertToPem($key['x5c'][0]);
+                if (isset($key['kid']) && $key['kid'] === $kid) {
+                    $r_key = $key;
+                    break;
+                }
             }
+        }
 
-            $this->cache->set($url, $secret);
+        // If a key was not found or the property does not exist, return.
+        if (is_null($r_key) || ! isset($r_key[$prop])) {
+            return null;
         }
 
-        return $secret;
+        // If the value is an array, get the first element.
+        return is_array( $r_key[$prop] ) ? $r_key[$prop][0] : $r_key[$prop];
```

Comment:
If I'm iterating over props of a file I know, I'd like to receive it as is, not being handled as non-array stuff. So I expect this method behaves the same on "any type of prop". e.g. make this return the array or null, and you later could do result[0] on the base of knowing how the JWKS is constructed.